hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java

```diff
@@ -16,247 +16,51 @@
 
 package com.hazelcast.query.impl;
 
-import com.hazelcast.core.TypeConverter;
 import com.hazelcast.internal.serialization.InternalSerializationService;
-import com.hazelcast.monitor.impl.IndexOperationStats;
 import com.hazelcast.monitor.impl.PerIndexStats;
-import com.hazelcast.nio.ObjectDataInput;
-import com.hazelcast.nio.ObjectDataOutput;
-import com.hazelcast.nio.serialization.Data;
-import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.query.impl.getters.Extractors;
-import com.hazelcast.query.impl.predicates.PredicateDataSerializerHook;
 
 import java.util.Collections;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
-import static com.hazelcast.query.impl.TypeConverters.NULL_CONVERTER;
-import static com.hazelcast.util.SetUtil.createHashSet;
-
-public class IndexImpl implements InternalIndex {
-
-    public static final NullObject NULL = new NullObject();
-
-    protected final InternalSerializationService ss;
-    protected final Extractors extractors;
-    protected final IndexStore indexStore;
-
-    private final String attributeName;
-    private final boolean ordered;
-    private final IndexCopyBehavior copyBehavior;
-    private final PerIndexStats stats;
+/**
+ * Provides implementation of on-heap indexes.
+ */
+public class IndexImpl extends AbstractIndex {
 
-    private volatile TypeConverter converter;
+    private final Set<Integer> indexedPartitions = Collections.newSetFromMap(new ConcurrentHashMap<Integer, Boolean>());
```

Comment:
can be a fixed size AtomicIntegerArray or something array based like that?

Reply:
Yes, it can, but there is no that much value in doing so: (1) we would waste `partition_count` - `partition_count` / `member_count` slots in the array and (2) execution paths (index addition and migration) touching the set are not performance critical.

Another approach is creating something like `Concurrent/AtomicBitSet`, so we may pack 271 partitions in 9 ints or 5 longs, but I don't think this task deserves such complexity :)

Reply:
ok makes sense